src/cunumeric/stat/histogram_omp.cc

```diff
@@ -0,0 +1,209 @@
+/* Copyright 2023 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#include "cunumeric/stat/histogram.h"
+#include "cunumeric/stat/histogram_template.inl"
+
+#define _USE_THRUST_
+
+#ifdef _USE_THRUST_
+#include "cunumeric/stat/histogram_cpu.h"
+#include "cunumeric/stat/histogram_impl.h"
+#endif
+
+#include <omp.h>
+
+#include <algorithm>
+#include <numeric>
+#include <tuple>
+
+namespace cunumeric {
+using namespace legate;
+
+namespace detail {
+
+// RO accessor (size, pointer) extractor:
+//
+template <typename VAL>
+std::tuple<size_t, const VAL*> get_accessor_ptr(const AccessorRO<VAL, 1>& src_acc,
+                                                const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+// RD accessor (size, pointer) extractor:
+//
+template <typename VAL>
+std::tuple<size_t, VAL*> get_accessor_ptr(const AccessorRD<SumReduction<VAL>, true, 1>& src_acc,
+                                          const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size = src_rect.hi - src_rect.lo + 1;
+  return std::make_tuple(src_size, src_ptr);
+}
+#ifdef _USE_THRUST_
+// accessor copy utility:
+//
+template <typename VAL>
+std::tuple<size_t, Buffer<VAL>, const VAL*> make_accessor_copy(const AccessorRO<VAL, 1>& src_acc,
+                                                               const Rect<1>& src_rect)
+{
+  size_t src_strides[1];
+  const VAL* src_ptr = src_acc.ptr(src_rect, src_strides);
+  assert(src_strides[0] == 1);
+  //
+  // const VAL* src_ptr: need to create a copy with create_buffer(...);
+  // since src will get sorted (in-place);
+  //
+  size_t src_size      = src_rect.hi - src_rect.lo + 1;
+  Buffer<VAL> src_copy = create_buffer<VAL>(src_size);
+  return std::make_tuple(src_size, src_copy, src_ptr);
+}
+#endif
+}  // namespace detail
+
+template <Type::Code CODE>
+struct HistogramImplBody<VariantKind::OMP, CODE> {
```

Comment:
I don't see how that could work, or any other example in the code where these 2 paths are coalesced into one TU. Can you point me to something similar done in the code?

Also, I'd like to keep these TU separate in case I have to fall back on the non-thrust OMP implementation (currently if-defed-out).

Reply:
> I don't see how that could work, or any other example in the code where these 2 paths are coalesced into one TU. Can you point me to something similar done in the code?

It's not exactly the same, but `matvecmul_cpu.inl` follows this pattern.

> Also, I'd like to keep these TU separate in case I have to fall back on the non-thrust OMP implementation (currently if-defed-out).

I think once we're done with testing on this branch, removing duplication is more valuable than the convenience of switching to an alternative implementation.

But in any case, you can have both. The thrust version of the code could be moved out to a common template in `histogram_cpu.h`:

```
template <exe_pol_t>
cpu_implementation(exe_pol_t exe_pol,
                   const AccessorRO<VAL, 1>& src,
                   const Rect<1>& src_rect,
                   const AccessorRO<BinType, 1>& bins,
                   const Rect<1>& bins_rect,
                   const AccessorRO<WeightType, 1>& weights,
                   const Rect<1>& weights_rect,
                   const AccessorRD<SumReduction<WeightType>, true, 1>& result,
                   const Rect<1>& result_rect);
```

then `histogram.cc` just does:

```
  void operator()(const AccessorRO<VAL, 1>& src,
                  const Rect<1>& src_rect,
                  const AccessorRO<BinType, 1>& bins,
                  const Rect<1>& bins_rect,
                  const AccessorRO<WeightType, 1>& weights,
                  const Rect<1>& weights_rect,
                  const AccessorRD<SumReduction<WeightType>, true, 1>& result,
                  const Rect<1>& result_rect) const
  {
    cpu_implementation(thrust::host, ...)
  }
```

and `histogram_omp.cc` does:

```
  void operator()(const AccessorRO<VAL, 1>& src,
                  const Rect<1>& src_rect,
                  const AccessorRO<BinType, 1>& bins,
                  const Rect<1>& bins_rect,
                  const AccessorRO<WeightType, 1>& weights,
                  const Rect<1>& weights_rect,
                  const AccessorRD<SumReduction<WeightType>, true, 1>& result,
                  const Rect<1>& result_rect) const
  {
#ifdef _USE_THRUST_
    cpu_implementation(thrust::omp::par, ...)
#else
    // alternative implementation
#endif
  }
```

Reply:
I'm inclined to refactor this in a follow-up PR, if that's okay.